### PR TITLE
Create single global lock, program wide, for the phase one of lock stabilisation.

### LINKED ISSUE #20684: Stabilize lock statement
Lang issue to track https://github.com/ballerina-platform/ballerina-spec/issues/357.

Design : https://github.com/ballerina-platform/ballerina-spec/commit/88daa72286a288c9722bdd7ca341888e189a1f6d

### PR DESCRIPTION
**Pull request by dulvinw**
## Purpose
With this PR, we remove per global variable locks, and use a program wide lock instance. 

Fixes #20684

## Check List 
- [ ] Read the [Contributing Guide](https://github.com/ballerina-platform/ballerina-lang/blob/master/CONTRIBUTING.md)
- [ ] Updated Change Log
- [ ] Checked Tooling Support (#<Issue Number>)
- [ ] Added necessary tests
  - [ ] Unit Tests
  - [ ] Spec Conformance Tests
  - [ ] Integration Tests
  - [ ] Ballerina By Example Tests
- [ ] Increased Test Coverage   
- [ ] Added necessary documentation  
  - [ ] API documentation 
  - [ ] Module documentation in Module.md files
  - [ ] Ballerina By Examples

### CODE REVIEW

**Comment by lankavitharana** [comment]
@dulvinw So with this, if a resource has a lock then that resource effectively works sequentially (not only single resource, if there are multiple resources with unrelated locks, even those cannot be accessed parallely), being an optimised language for network development, I feel this kind of betrays that aspect, hope there is some solution for that as well?

**Comment by dulvinw** [comment]
@lankavitharana Yes, this is suppose to be the initial implementation. Then we will follow through to infer the locks by analyzing the lock-blocks

File: bvm/ballerina-runtime/src/main/java/org/ballerinalang/jvm/BLockManager.java [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+ */
+public class BLockManager {
+
+    /**
+     * The map of locks inferred.
+     */
+    private static Map<String, BLock> globalLockMap;
+
+    static {
+        globalLockMap = new HashMap<>();
```
Comment (rdhananjaya) [inline bvm/ballerina-runtime/src/main/java/org/ballerinalang/jvm/BLockManager.java]: I think we can use a concurrent hashmap here, and get rid of manual synchronization.
https://docs.oracle.com/javase/8/docs/api/java/util/concurrent/ConcurrentHashMap.html

File: bvm/ballerina-runtime/src/main/java/org/ballerinalang/jvm/BLockManager.java [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+package org.ballerinalang.jvm;
+
+import java.util.HashMap;
+import java.util.Map;
+
+/**
+ *
+ * @since 1.2.0
+ */
+public class BLockManager {
```
Comment (rdhananjaya) [inline bvm/ballerina-runtime/src/main/java/org/ballerinalang/jvm/BLockManager.java]: This looks more like a lock store/registry, shall we add a proper name reflecting the expected behavior.

Comment (dulvinw) [reply bvm/ballerina-runtime/src/main/java/org/ballerinalang/jvm/BLockManager.java]: Sure. Is BLockStore a better name?

Comment (rdhananjaya) [reply bvm/ballerina-runtime/src/main/java/org/ballerinalang/jvm/BLockManager.java]: +1

File: bvm/ballerina-runtime/src/main/java/org/ballerinalang/jvm/BLockManager.java (hunk 1 above)
Comment (dulvinw) [reply bvm/ballerina-runtime/src/main/java/org/ballerinalang/jvm/BLockManager.java]: +1

File: bvm/ballerina-runtime/src/main/java/org/ballerinalang/jvm/BLockStore.java [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+
+    public BLockStore() {
+        globalLockMap = new ConcurrentHashMap<>();
+    }
+
+    public void addLockToMap(String lockName) {
+        globalLockMap.put(lockName, new BLock());
+    }
+
+    public synchronized BLock getLockFromMap(String lockName) {
```
Comment (KRVPerera) [inline bvm/ballerina-runtime/src/main/java/org/ballerinalang/jvm/BLockStore.java]: I think synchronized does not support your intention here. ```computeIfAbsent``` is atomic according to [docs](https://docs.oracle.com/javase/8/docs/api/java/util/concurrent/ConcurrentHashMap.html#computeIfAbsent-K-java.util.function.Function-).
If the only getter we have is creating a lock too maybe we can remove ```addLockToMap``` function

File: bvm/ballerina-runtime/src/main/java/org/ballerinalang/jvm/BLockStore.java [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+ */
+public class BLockStore {
+
+    /**
+     * The map of locks inferred.
+     */
+    private  Map<String, BLock> globalLockMap;
+
+    public BLockStore() {
+        globalLockMap = new ConcurrentHashMap<>();
```
Comment (KRVPerera) [inline bvm/ballerina-runtime/src/main/java/org/ballerinalang/jvm/BLockStore.java]: Shouldn't this be a thread sage singleton

Comment (dulvinw) [reply bvm/ballerina-runtime/src/main/java/org/ballerinalang/jvm/BLockStore.java]: This class is accessed only once per program, using the init class in jbal

File: bvm/ballerina-runtime/src/main/java/org/ballerinalang/jvm/BLockStore.java (hunk 3 above)
Comment (dulvinw) [reply bvm/ballerina-runtime/src/main/java/org/ballerinalang/jvm/BLockStore.java]: Yes, should remove this.

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/tree/statements/BLangLock.java [hunk 5]
```diff
@@ -43,10 +32,6 @@
 
     public BLangBlockStmt body;
 
-    public Set<BVarSymbol> lockVariables = new HashSet<>();
-
-    public Map<BVarSymbol, Set<BLangStructFieldAccessExpr>> fieldVariables = new HashMap<>();
-
     public String uuid;
```
Comment (hasithaa) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/tree/statements/BLangLock.java]: this is an unused variable.

**Pull request merged**